README.md
# blake2

[![pub package](https://img.shields.io/pub/v/blake2.svg)](https://pub.dartlang.org/packages/blake2)

Pure Dart implementations of the BLAKE2b and BLAKE2s cryptographic
hashing functions.

# Usage

```dart
import 'package:blake2/blake2.dart';
```

The blake2 library contains 3 classes:
[Blake2b], [blake2s], and their parent class: [blake2].

Both [Blake2b] and [Blake2s] can be instanced directly, accepting
[Uint8List]s for the [key], [salt] and [personalization] parameters.

An [iv] can also be specified as a [Uint64List] or [Uint32List],
for [Blake2b] and [Blake2s] respectively.

All of the parameters are optional and can be left null.

The hash can be returned as a [Uint8List] with the `digest()` method,
or as a [String] with the `digestToString()` method.

```dart
const String key = '7a546d0acf4f30d371c505f32548c867';
const String salt = '39b69017';
const String personalization = '4d97847f';

final Blake2b blake2b = Blake2b(
  key: Uint8List.fromList(key.codeUnits),
  salt: Uint8List.fromList(salt.codeUnits),
  personalization: Uint8List.fromList(personalization.codeUnits),
);

print(blake2b.digest());

// Prints: [179, 80, 231, 226, 92, 124, 89, 99, 133, 217, 254, 167, 78, 189, 21, 254, 209, 126, 124, 142, 180, 193, 37, 196, 83, 167, 106, 44, 107, 178, 135, 23]

print(blake2b.digestToString());

// Prints: ÓZê]×8£ÛU]ÚÙ¬DÛ!;s¶Ø

blake2b.update(Uint8List.fromList([110, 101, 119, 32, 100, 97, 116, 97]));

print(blake2b.digestToString());

// Prints: {é2¾ë];²à&Õ½BXN¼ý(³ö(î&?½ND_
```

Each class also has a static method `fromStrings()`, that accepts
the [key], [salt], and [personalization] parameters as [String]s, and
a method `updateWithString()` to add additional data as a [String].

```dart
final Blake2s blake2s = Blake2s.fromStrings(
  key: '2805f1ad825b3489dae4c5526014c1ed',
  salt: '8d6efe2c',
  personalization: 'e9f6def7',
);

print(blake2s.digestToString());

// Prints: ÃÉØNPnû·Ì÷Eæüç-SºØ·Ò=ëik

blake2s.updateWithString('new data');

print(blake2s.digestToString());

// Prints: ßlF©óþ;+®Ð¸BaðÅª_#Y¥\²p

```


CHANGELOG.md
## [0.1.2] - May 25, 2022

* Updated to null-safety

## [0.1.1] - January 15, 2020

* Formatting changes

## [0.1.0] - July 27, 2019

* Initial release


test/blake2_test.dart
import 'dart:io';
import 'dart:typed_data';

import 'package:blake2/blake2.dart';
import 'package:test/test.dart';

final List<String?> keys = _getVectors('keys.txt');
final List<String?> salts = _getVectors('salts.txt');
final List<String?> personalizations = _getVectors('personalizations.txt');
final List<String?> updates = _getVectors('updates.txt');

void main() {
  test('BLAKE2b Hashes', () {
    final hashes = _calculateHashes(Blake2b);

    expect(_compareHashes(hashes, 'blake2b.txt'), equals(true));
  });

  test('BLAKE2s Hashes', () {
    final hashes = _calculateHashes(Blake2s);

    expect(_compareHashes(hashes, 'blake2s.txt'), equals(true));
  });
}

List<Uint8List> _calculateHashes(Type type) {
  assert((type == Blake2b || type == Blake2s));

  final hashes = List<Uint8List?>.filled(keys.length, null);

  var updateCount = 0;

  for (var i = 0; i < keys.length; i++) {
    final numUpdates = i % 4;

    final blake2 = (type == Blake2b)
        ? Blake2b.fromStrings(
            key: keys[i],
            salt: salts[i],
            personalization: personalizations[i],
          )
        : Blake2s.fromStrings(
            key: keys[i],
            salt: salts[i],
            personalization: personalizations[i],
          );

    for (var j = 0; j < numUpdates; j++) {
      blake2.updateWithString(updates[updateCount]!);
      updateCount++;
    }

    var hash = blake2.digest();

    hashes[i] = hash;
  }

  return hashes.cast();
}

List<String?> _getVectors(String filename) {
  final file = File('test/test_vectors/$filename');
  final vectors = file.readAsLinesSync();

  return vectors.map((e) => e.isEmpty ? null : e).toList();
}

bool _compareHashes(List<Uint8List> tests, String filename) {
  final vectors = _getVectors('hashes/$filename');

  expect(vectors.length, tests.length);

  for (var i = 0; i < vectors.length; i++) {
    final vector = Uint8List.fromList(
      vectors[i]!.split(',').map(int.parse).toList(),
    );

    for (var j = 0; j < vector.length; j++) {
      if (vector[j] != tests[i][j]) return false;
    }
  }

  return true;
}


lib/src/blake2b.dart
import 'dart:typed_data';
import './blake2.dart';

/// The BLAKE2b (64-bit) flavor of the BLAKE2
/// cryptographic hash function.
class Blake2b extends Blake2 {
  /// The BLAKE2s (32-bit) flavor of the BLAKE2
  /// cryptographic hash function.
  ///
  /// [digestLength] must not be null and must be `> 0`.
  ///
  /// [key] may be null, but if set, it will be used for the
  /// first round of compression.
  ///
  /// [salt], [personalization], and [iv] may be null
  /// or must `== 8` characters in length.
  Blake2b({
    this.digestLength = 32,
    this.key,
    this.salt,
    this.personalization,
    this.iv,
  })  : assert( digestLength > 0 && digestLength <= 32),
        assert(salt == null || salt.length == 8),
        assert(personalization == null || personalization.length == 8),
        assert(iv == null || iv.length == 8) {
    iv ??= Uint64List.fromList(<int>[
      0x6a09e667f3bcc908, 0xbb67ae8584caa73b,
      0x3c6ef372fe94f82b, 0xa54ff53a5f1d36f1,
      0x510e527fade682d1, 0x9b05688c2b3e6c1f,
      0x1f83d9abfb41bd6b, 0x5be0cd19137e2179,
    ]);

    reset();
  }

  @override
  final Uint8List? key;

  @override
  final Uint8List? salt;

  @override
  final Uint8List? personalization;

  @override
  final int digestLength;

  @override
  Uint64List? iv;

  @override
  final int bitLength = 64;

  @override
  final Uint8List sigma = Uint8List.fromList([
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
    14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3,
    11, 8, 12, 0, 5, 2, 15, 13, 10, 14, 3, 6, 7, 1, 9, 4,
    7, 9, 3, 1, 13, 12, 11, 14, 2, 6, 5, 10, 4, 0, 15, 8,
    9, 0, 5, 7, 2, 4, 10, 15, 14, 1, 11, 12, 6, 8, 3, 13,
    2, 12, 6, 10, 0, 11, 8, 3, 4, 13, 7, 5, 15, 14, 1, 9,
    12, 5, 1, 15, 14, 13, 4, 10, 0, 7, 6, 3, 9, 2, 8, 11,
    13, 11, 7, 14, 12, 1, 3, 9, 5, 0, 15, 4, 8, 6, 2, 10,
    6, 15, 14, 9, 11, 3, 0, 8, 12, 2, 13, 7, 1, 4, 10, 5,
    10, 2, 8, 4, 7, 6, 1, 5, 15, 11, 9, 14, 3, 12, 13, 0,
    0, 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15,
    14, 10, 4, 8, 9, 15, 13, 6, 1, 12, 0, 2, 11, 7, 5, 3,
  ]);

  /// Returns a [Blake2b] instance using [Strings] for the
  /// [key], [salt], and [personalization] arguments.
  ///
  /// [key] may be null and can be up to 32 characters long.
  ///
  /// [salt] and [personalization] may be null and
  /// must both be 8 characters in length.
  static Blake2b fromStrings({
    int digestLength = 32,
    String? key,
    String? salt,
    String? personalization,
    Uint64List? iv,
  }) {
    assert( digestLength > 0 && digestLength <= 32);
    assert(salt == null || salt.length == 8);
    assert(personalization == null || personalization.length == 8);
    assert(iv == null || iv.length == 8);

    return Blake2b(
      digestLength: digestLength,
      key: (key == null) ? key as Uint8List? : Uint8List.fromList(key.codeUnits),
      salt: (salt == null) ? salt as Uint8List? : Uint8List.fromList(salt.codeUnits),
      personalization: (personalization != null)
          ? Uint8List.fromList(personalization.codeUnits)
          : null,
      iv: iv,
    );
  }
}


lib/src/blake2s.dart
import 'dart:typed_data';
import './blake2.dart';

/// The BLAKE2s (32-bit) flavor of the BLAKE2
/// cryptographic hash function.
class Blake2s extends Blake2 {
  /// The BLAKE2s (32-bit) flavor of the BLAKE2
  /// cryptographic hash function.
  ///
  /// [digestLength] must not be null and must be `> 0`.
  ///
  /// [key] may be null, but if set, it will be used for the
  /// first round of compression.
  ///
  /// [salt], [personalization], and [iv] may be null
  /// or must `== 8` characters in length.
  Blake2s({
    this.digestLength = 32,
    this.key,
    this.salt,
    this.personalization,
    this.iv,
  })  : assert(digestLength > 0 && digestLength <= 32),
        assert(salt == null || salt.length == 8),
        assert(personalization == null || personalization.length == 8),
        assert(iv == null || iv.length == 8) {
    iv ??= Uint32List.fromList(<int>[
      0x6A09E667,
     0xBB67AE85,
     0x3C6EF372,
     0xA54FF53A,
      0x510E527F,
     0x9B05688C,
     0x1F83D9AB,
     0x5BE0CD19,
    ]);

    reset();
  }

  @override
  final Uint8List? key;

  @override
  final Uint8List? salt;

  @override
  final Uint8List? personalization;

  @override
  final int digestLength;

  @override
  Uint32List? iv;

  @override
  final int bitLength = 32;

  @override
  final Uint8List sigma = Uint8List.fromList(<int>[
    0,
   1,
   2,
   3,
   4,
   5,
   6,
   7,
   8,
   9,
   10,
   11,
   12,
   13,
   14,
   15,
    14,
   10,
   4,
   8,
   9,
   15,
   13,
   6,
   1,
   12,
   0,
   2,
   11,
   7,
   5,
   3,
    11,
   8,
   12,
   0,
   5,
   2,
   15,
   13,
   10,
   14,
   3,
   6,
   7,
   1,
   9,
   4,
    7,
   9,
   3,
   1,
   13,
   12,
   11,
   14,
   2,
   6,
   5,
   10,
   4,
   0,
   15,
   8,
    9,
   0,
   5,
   7,
   2,
   4,
   10,
   15,
   14,
   1,
   11,
   12,
   6,
   8,
   3,
   13,
    2,
   12,
   6,
   10,
   0,
   11,
   8,
   3,
   4,
   13,
   7,
   5,
   15,
   14,
   1,
   9,
    12,
   5,
   1,
   15,
   14,
   13,
   4,
   10,
   0,
   7,
   6,
   3,
   9,
   2,
   8,
   11,
    13,
   11,
   7,
   14,
   12,
   1,
   3,
   9,
   5,
   0,
   15,
   4,
   8,
   6,
   2,
   10,
    6,
   15,
   14,
   9,
   11,
   3,
   0,
   8,
   12,
   2,
   13,
   7,
   1,
   4,
   10,
   5,
    10,
   2,
   8,
   4,
   7,
   6,
   1,
   5,
   15,
   11,
   9,
   14,
   3,
   12,
   13,
   0,
  ]);

  /// Returns a [Blake2s] instance using [Strings] for the
  /// [key], [salt], and [personalization] arguments.
  ///
  /// [key] may be null and can be up to 32 characters long.
  ///
  /// [salt] and [personalization] may be null and
  /// must both be 8 characters in length.
  static Blake2s fromStrings({
    int digestLength = 32,
    String? key,
    String? salt,
    String? personalization,
    Uint32List? iv,
  }) {
    assert(digestLength > 0 && digestLength <= 32);
    assert(salt == null || salt.length == 8);
    assert(personalization == null || personalization.length == 8);
    assert(iv == null || iv.length == 8);

    return Blake2s(
      digestLength: digestLength,
      key: (key == null) ? null : Uint8List.fromList(key.codeUnits),
      salt: (salt == null) ? null : Uint8List.fromList(salt.codeUnits),
      personalization: (personalization != null)
          ? Uint8List.fromList(personalization.codeUnits)
          : null,
      iv: iv,
    );
  }
}


lib/src/blake2.dart
import 'dart:typed_data';

/// The base class containing the shared values and methods of the
/// [Blake2b] and [Blake2s] hashing algorithms.
abstract class Blake2 {
  /// The length of the digest, defaults to `32`.
  int get digestLength;

  /// If set, [key] is used for the first round of compression.
  Uint8List? get key;

  /// If set, [salt] is used to modify the initialization vector.
  Uint8List? get salt;

  /// If set, [personalization] acts as a second [salt].
  Uint8List? get personalization;

  /// Initialization vector
  List<int>? get iv;

  /// Offsets for each round within the memory block.
  Uint8List get sigma;

  /// The hash of the key + all of the values added via `update()`.
  late List<int> _hash;

  /// The bit-length of the integers being used in the hashing function.
  ///
  /// BLAKE2b uses 64-bit, and BLAKE2s uses 32-bit.
  int get bitLength;

  /// The buffer block's length
  int get _blockSize => bitLength * 2;

  /// Buffer block
  late List<int> _block;

  /// Resets the hash to its initial state, effectively
  /// clearing all values added via `update()`.
  void reset() {
    final keyLength = (key == null) ? 0 : key!.length;

    final hash = List<int>.filled(8, 0);

    for (var i = 0; i < hash.length; i++) {
      hash[i] = iv![i];
      if (salt != null) hash[i] ^= salt![i];
      if (personalization != null) hash[i] ^= personalization![i];
    }

    final block = List<int>.filled(_blockSize, 0);

    if (bitLength == 32) {
      _hash = Uint32List.fromList(hash);
      _block = Uint32List.fromList(block);
    } else {
      _hash = Uint64List.fromList(hash);
      _block = Uint64List.fromList(block);
    }

    // If [key] exists, make the first round with it.
    if (keyLength > 0) {
      update(key!);
      _pointer = _blockSize;
    }
  }

  int _pointer = 0;

  int _counter = 0;

  /// Calculates the digest of all data passed via `update()`.
  Uint8List digest() {
    final digest = Uint8List(digestLength);

    _counter += _pointer;

    // Clear block
    while (_pointer < _blockSize) {
      _block[_pointer++] = 0;
    }

    // Compress
    _compress(true);

    // Little-endian convert and store
    for (var i = 0; i < digestLength; i++) {
      digest[i] = _hash[i >> 2] >> (8 * (i & 3));
    }

    return digest;
  }

  /// Returns the calculated digest as a string.
  String digestToString() => String.fromCharCodes(digest());

  /// Update hash content with the given data.
  void update(Uint8List? data) {
    assert(data != null);

    for (var i = 0; i < data!.length; i++) {
      if (_pointer == _blockSize) {
        _counter += _pointer;
        _compress(false);
        _pointer = 0;
      }

      // Copy input array to input block.
      _block[_pointer++] = data[i];
    }
  }

  /// Converts [data] to a [Uint8List] and passes it to `update()`.
  void updateWithString(String? data) {
    assert(data != null);

    update(Uint8List.fromList(data!.codeUnits));
  }

  /// Compression function
  void _compress(bool isLast) {
    List<int> v;
    List<int> m;

    final vm = List<int>.filled(16, 0);

    if (bitLength == 32) {
      v = Uint32List.fromList(vm);
      m = Uint32List.fromList(vm);
    } else {
      v = Uint64List.fromList(vm);
      m = Uint64List.fromList(vm);
    }

    for (var i = 0; i < 8; i++) {
      v[i] = _hash[i];
      v[i + 8] = iv![i];
    }

    v[12] ^= _counter;
    v[13] ^= _counter;

    if (isLast) {
      v[14] ^= ~v[14];
    }

    // Get Little-endian words
    for (var i = 0; i < 16; i++) {
      m[i] = _get32(_block, i * 4);
    }

    void _gamma(int a, int b, int c, int d, int x, int y) {
      v[a] += v[b] + x;
      v[d] = _rotateRight(v[d] ^ v[a], 16, bitLength);

      v[c] += v[d];
      v[b] = _rotateRight(v[b] ^ v[c], 12, bitLength);

      v[a] += v[b] + y;
      v[d] = _rotateRight(v[d] ^ v[a], 8, bitLength);

      v[c] += v[d];
      v[b] = _rotateRight(v[b] ^ v[c], 7, bitLength);
    }

    final compressionRounds = (bitLength == 32) ? 10 : 12;

    for (var i = 0; i < compressionRounds; i++) {
      _gamma(0, 4, 8, 12, m[sigma[i * 16]], m[sigma[i * 16 + 1]]);
      _gamma(1, 5, 9, 13, m[sigma[i * 16 + 2]], m[sigma[i * 16 + 3]]);
      _gamma(2, 6, 10, 14, m[sigma[i * 16 + 4]], m[sigma[i * 16 + 5]]);
      _gamma(3, 7, 11, 15, m[sigma[i * 16 + 6]], m[sigma[i * 16 + 7]]);
      _gamma(0, 5, 10, 15, m[sigma[i * 16 + 8]], m[sigma[i * 16 + 9]]);
      _gamma(1, 6, 11, 12, m[sigma[i * 16 + 10]], m[sigma[i * 16 + 11]]);
      _gamma(2, 7, 8, 13, m[sigma[i * 16 + 12]], m[sigma[i * 16 + 13]]);
      _gamma(3, 4, 9, 14, m[sigma[i * 16 + 14]], m[sigma[i * 16 + 15]]);
    }

    for (var i = 0; i < 8; i++) {
      _hash[i] ^= v[i] ^ v[i + 8];
    }
  }

  /// Convert 4 bytes to Little-endian word.
  int _get32(List<int> data, int index) {

    return data[index++] ^
        (data[index++] << 8) ^
        (data[index++] << 16) ^
        (data[index] << 24);
  }

  /// Cyclic right rotation
  int _rotateRight(int data, int shift, int length) {

    return (data >> shift) ^ (data << (length - shift));
  }
}
